tests/alpn_handler_test.c

```diff
@@ -0,0 +1,316 @@
+    struct aws_channel_slot *new_slot,
+    struct aws_byte_buf *protocol,
+    void *ctx) {
+    struct alpn_test_on_negotiation_args *negotiation_args = (struct alpn_test_on_negotiation_args *)ctx;
+
+    struct aws_channel_handler *handler =
+        (struct aws_channel_handler *)aws_mem_acquire(negotiation_args->allocator, sizeof(struct aws_channel_handler));
+
+    negotiation_args->new_handler = handler;
+    negotiation_args->protocol = *protocol;
+    negotiation_args->new_slot = new_slot;
+
+    handler->vtable = &s_alpn_test_vtable;
+    handler->alloc = negotiation_args->allocator;
+
+    return handler;
+}
+
+static bool s_alpn_test_shutdown_predicate(void *arg) {
+    struct alpn_channel_setup_test_args *test_args = (struct alpn_channel_setup_test_args *)arg;
+    return test_args->shutdown_finished;
+}
+
+static void s_on_server_channel_on_shutdown(struct aws_channel *channel, int error_code, void *user_data) {
+
+    (void)channel;
+    (void)error_code;
+
+    struct alpn_channel_setup_test_args *test_args = (struct alpn_channel_setup_test_args *)user_data;
+
+    test_args->shutdown_finished = true;
+    aws_condition_variable_notify_one(&test_args->condition_variable);
+}
+
+static int s_test_alpn_successfully_negotiates(struct aws_allocator *allocator, void *ctx) {
+
+    (void)ctx;
+
+    struct aws_event_loop *event_loop = aws_event_loop_new_default(allocator, aws_high_res_clock_get_ticks);
+
+    ASSERT_NOT_NULL(event_loop, "Event loop creation failed with error: %s", aws_error_debug_str(aws_last_error()));
+    ASSERT_SUCCESS(aws_event_loop_run(event_loop));
+    struct aws_channel channel;
+
+    struct alpn_channel_setup_test_args test_args = {
+        .error_code = 0, .condition_variable = AWS_CONDITION_VARIABLE_INIT, .shutdown_finished = false};
+
+    struct aws_channel_creation_callbacks callbacks = {
+        .on_setup_completed = s_alpn_channel_setup_test_on_setup_completed,
+        .setup_user_data = &test_args,
+        .on_shutdown_completed = s_on_server_channel_on_shutdown,
+        .shutdown_user_data = &test_args,
+    };
+
+    struct aws_mutex mutex = AWS_MUTEX_INIT;
+
+    ASSERT_SUCCESS(aws_mutex_lock(&mutex));
+    ASSERT_SUCCESS(aws_channel_init(&channel, allocator, event_loop, &callbacks));
+    ASSERT_SUCCESS(aws_condition_variable_wait(&test_args.condition_variable, &mutex));
+
+    struct aws_channel_slot *slot = aws_channel_slot_new(&channel);
+    ASSERT_NOT_NULL(slot);
+
+    struct alpn_test_on_negotiation_args on_negotiation_args = {
+        .new_slot = NULL, .protocol = {0}, .new_handler = NULL, .allocator = allocator};
+
+    struct aws_channel_handler *handler =
+        aws_tls_alpn_handler_new(allocator, s_alpn_tls_successful_negotiation, &on_negotiation_args);
+    ASSERT_NOT_NULL(handler);
+    ASSERT_SUCCESS(aws_channel_slot_set_handler(slot, handler));
+
+    struct aws_tls_negotiated_protocol_message protocol_message = {.protocol = aws_byte_buf_from_c_str("h2")};
```

Comment:
Should we add a separate test for a list of protocols?

Reply:
there's no universe where you'd get a list of protocols after a TLS negotiation.